### 1) Имеются следующие данные о транзакциях в период с 01.12.2010 по 12.09.2011:
InvoiceNo – номер транзакции
StockCode – код товара
Description – описание товара
Quantity – количество единиц товара, добавленных в заказ
InvoiceDate – дата транзакции 
UnitPrice – цена за единицу товара
CustomerID – id клиента
Country – страна, где проживает клиент
Импортируйте данные data.csv.zip (в папке) с кодировкой ISO-8859-1. Запишите полученный датафрейм в retail, а названия колонок сохраните в переменную retail_columns.

In [121]:
import pandas as pd

In [122]:
retail = pd.read_csv('C:/Users/Eternity/data.csv.zip', compression='zip', encoding='ISO-8859-1')
retail_columns = retail.columns

In [123]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [124]:
retail.shape

(541909, 8)

### 2) Проверьте, встречаются ли в данных повторяющиеся наблюдения, и в качестве ответа укажите их количество. Если они есть, то удалите их из retail.

In [125]:
retail[retail.duplicated()]

InvoiceNo StockCode                        Description  Quantity  \
517       536409     21866        UNION JACK FLAG LUGGAGE TAG         1   
527       536409     22866      HAND WARMER SCOTTY DOG DESIGN         1   
537       536409     22900    SET 2 TEA TOWELS I LOVE LONDON          1   
539       536409     22111       SCOTTIE DOG HOT WATER BOTTLE         1   
555       536412     22327  ROUND SNACK BOXES SET OF 4 SKULLS         1   
...          ...       ...                                ...       ...   
541675    581538     22068        BLACK PIRATE TREASURE CHEST         1   
541689    581538     23318     BOX OF 6 MINI VINTAGE CRACKERS         1   
541692    581538     22992             REVOLVER WOODEN RULER          1   
541699    581538     22694                       WICKER STAR          1   
541701    581538     23343       JUMBO BAG VINTAGE CHRISTMAS          1   

            InvoiceDate  UnitPrice  CustomerID         Country  
517     12/1/2010 11:45       1.25     17908.0  United Kingdom  
527     12/1/2010 11:45       2.10     17908.0  United Kingdom  
537     12/1/2010 11:45       2.95     17908.0  United Kingdom  
539     12/1/2010 11:45       4.95     17908.0  United Kingdom  
555     12/1/2010 11:49       2.95     17920.0  United Kingdom  
...                 ...        ...         ...             ...  
541675  12/9/2011 11:34       0.39     14446.0  United Kingdom  
541689  12/9/2011 11:34       2.49     14446.0  United Kingdom  
541692  12/9/2011 11:34       1.95     14446.0  United Kingdom  
541699  12/9/2011 11:34       2.10     14446.0  United Kingdom  
541701  12/9/2011 11:34       2.08     14446.0  United Kingdom  

[5268 rows x 8 columns]

In [126]:
retail[retail.duplicated()].shape[0]

5268

In [127]:
retail = retail.drop_duplicates()

In [128]:
retail.shape[0]

536641

### 3) ⭐️Задание со звёздочкой!⭐️
Данные содержат в себе записи как и об успешных транзакциях, так и об отмененных. Если пользователь отменил заказ, в начале номера транзакции (InvoiceNo) ставится C (canceled). 
Сколько всего заказов отменили пользователи?

In [129]:
retail[retail['InvoiceNo'].str.startswith('C')]

InvoiceNo StockCode                       Description  Quantity  \
141      C536379         D                          Discount        -1   
154      C536383    35004C   SET OF 3 COLOURED  FLYING DUCKS        -1   
235      C536391     22556    PLASTERS IN TIN CIRCUS PARADE        -12   
236      C536391     21984  PACK OF 12 PINK PAISLEY TISSUES        -24   
237      C536391     21983  PACK OF 12 BLUE PAISLEY TISSUES        -24   
...          ...       ...                               ...       ...   
540449   C581490     23144   ZINC T-LIGHT HOLDER STARS SMALL       -11   
541541   C581499         M                            Manual        -1   
541715   C581568     21258        VICTORIAN SEWING BOX LARGE        -5   
541716   C581569     84978  HANGING HEART JAR T-LIGHT HOLDER        -1   
541717   C581569     20979     36 PENCILS TUBE RED RETROSPOT        -5   

            InvoiceDate  UnitPrice  CustomerID         Country  
141      12/1/2010 9:41      27.50     14527.0  United Kingdom  
154      12/1/2010 9:49       4.65     15311.0  United Kingdom  
235     12/1/2010 10:24       1.65     17548.0  United Kingdom  
236     12/1/2010 10:24       0.29     17548.0  United Kingdom  
237     12/1/2010 10:24       0.29     17548.0  United Kingdom  
...                 ...        ...         ...             ...  
540449   12/9/2011 9:57       0.83     14397.0  United Kingdom  
541541  12/9/2011 10:28     224.69     15498.0  United Kingdom  
541715  12/9/2011 11:57      10.95     15311.0  United Kingdom  
541716  12/9/2011 11:58       1.25     17315.0  United Kingdom  
541717  12/9/2011 11:58       1.25     17315.0  United Kingdom  

[9251 rows x 8 columns]

In [130]:
retail[retail['InvoiceNo'].str.startswith('C')].shape[0]

9251

### 4) Теперь отфильтруйте данные и оставьте в retail только те заказы, где Quantity > 0. В качестве ответа укажите число оставшихся строк.

In [131]:
retail[retail['Quantity'] > 0].shape[0]

526054

### 5) 💀 Сложное задание! 💀
Посчитайте число заказов для каждого пользователя (CustomerID) из Германии (Germany). Оставьте только тех, кто совершил более N транзакций (InvoiceNo), где N – 80% процентиль. Запишите полученные id пользователей в germany_top (не весь датафрейм, только id). Идентификатор заказа – InvoiceNo. Для каждого заказа в данных может встречаться более 1 строки.

In [132]:
germany_top = retail.query('Country == "Germany"') \
                    .groupby('CustomerID') \
                    .agg({'InvoiceNo': pd.Series.nunique}) \
                    .rename(columns={'InvoiceNo': 'count'}) \
                    .sort_values('count', ascending=False)

In [133]:
germany_top

count
CustomerID       
12471.0        49
12569.0        35
12474.0        30
12720.0        29
12709.0        26
...           ...
12622.0         1
12623.0         1
12644.0         1
12648.0         1
12605.0         1

[95 rows x 1 columns]

In [134]:
quantile_85 = germany_top.quantile(q=.85)

In [135]:
quantile_85

count    12.9
Name: 0.85, dtype: float64

In [136]:
germany_top_bayers = germany_top.query('count > 12.9').index

### 6) ⭐️Задание со звёздочкой!⭐️
Используя объект с id пользователей (germany_top), полученный на предыдущем шаге, отфильтруйте наблюдения и оставьте в данных записи только по интересующим нас юзерам. Результирующий датафрейм запишите в top_retail_germany.

In [137]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [138]:
top_retail_germany = retail.query('CustomerID in @germany_top_bayers')

In [139]:
top_retail_germany

InvoiceNo StockCode                          Description  Quantity  \
1973     C536548     22244           3 HOOK HANGER MAGIC GARDEN        -4   
1974     C536548     22242        5 HOOK HANGER MAGIC TOADSTOOL        -5   
1975     C536548     20914  SET/5 RED RETROSPOT LID GLASS BOWLS        -1   
1976     C536548     22892    SET OF SALT AND PEPPER TOADSTOOLS        -7   
1977     C536548     22654                   DELUXE SEWING KIT         -1   
...          ...       ...                                  ...       ...   
538563   C581409     82484    WOOD BLACK BOARD ANT WHITE FINISH        -1   
538564   C581409     82482    WOODEN PICTURE FRAME WHITE FINISH        -1   
538565   C581409     22173   METAL 4 HOOK HANGER FRENCH CHATEAU        -2   
538566   C581409    85199L  LARGE HANGING IVORY & RED WOOD BIRD        -1   
538567   C581409     85127   SMALL SQUARE CUT GLASS CANDLESTICK        -5   

            InvoiceDate  UnitPrice  CustomerID  Country  
1973    12/1/2010 14:33       1.95     12472.0  Germany  
1974    12/1/2010 14:33       1.65     12472.0  Germany  
1975    12/1/2010 14:33       2.95     12472.0  Germany  
1976    12/1/2010 14:33       1.25     12472.0  Germany  
1977    12/1/2010 14:33       5.95     12472.0  Germany  
...                 ...        ...         ...      ...  
538563  12/8/2011 14:08       7.95     12476.0  Germany  
538564  12/8/2011 14:08       2.95     12476.0  Germany  
538565  12/8/2011 14:08       3.29     12476.0  Germany  
538566  12/8/2011 14:08       0.65     12476.0  Germany  
538567  12/8/2011 14:08       4.95     12476.0  Germany  

[4179 rows x 8 columns]

### 7) Сгруппируйте top_retail_germany по коду товара (StockCode). Какой из продуктов добавляли в корзину чаще всего, кроме POST? 
Note: одним заказом считается единовременная покупка любого количества товара, т.е. без учета Quantity.

In [140]:
top_retail_germany.StockCode.value_counts()

POST      190
22423      55
22326      52
22328      38
22467      37
         ... 
23066       1
22821       1
21291       1
16168M      1
23562       1
Name: StockCode, Length: 1116, dtype: int64

### 8) Вернемся к анализу полного датасета retail. Создайте колонку Revenue с суммой покупки, используя колонки Quantity и UnitPrice.

In [141]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom

In [142]:
retail['Revenue'] = retail['Quantity'] * retail['UnitPrice']

In [143]:
retail['Revenue']

0         15.30
1         20.34
2         22.00
3         20.34
4         20.34
          ...  
541904    10.20
541905    12.60
541906    16.60
541907    16.60
541908    14.85
Name: Revenue, Length: 536641, dtype: float64

### 9) Для каждой транзакции (InvoiceNo), посчитайте финальную сумму заказа. В качестве ответа укажите топ-5 (InvoiceNo) по сумме заказа (через запятую с пробелом и в порядке убывания TotalRevenue)

Например, для следующего примера...

InvoiceNo StockCode Quantity InvoiceDate          UnitPrice  CustomerID  Revenue
536365    85123A    6        2010-12-01 08:26:00  2.55       17850.0     15.30
536365    71053     6        2010-12-01 08:26:00  3.39       17850.0     20.34
555555    71053     1        2010-12-02 18:00:00  3.39       17850.0     3.39
...сумма заказа будет равна:

InvoiceNo  TotalRevenue
536365     35.64
555555     3.39

И ответом будет:
536365, 555555

In [144]:
retail.head()

InvoiceNo StockCode                          Description  Quantity  \
0    536365    85123A   WHITE HANGING HEART T-LIGHT HOLDER         6   
1    536365     71053                  WHITE METAL LANTERN         6   
2    536365    84406B       CREAM CUPID HEARTS COAT HANGER         8   
3    536365    84029G  KNITTED UNION FLAG HOT WATER BOTTLE         6   
4    536365    84029E       RED WOOLLY HOTTIE WHITE HEART.         6   

      InvoiceDate  UnitPrice  CustomerID         Country  Revenue  
0  12/1/2010 8:26       2.55     17850.0  United Kingdom    15.30  
1  12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
2  12/1/2010 8:26       2.75     17850.0  United Kingdom    22.00  
3  12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34  
4  12/1/2010 8:26       3.39     17850.0  United Kingdom    20.34

In [153]:
retail.groupby('InvoiceNo', as_index=False).agg({'Revenue':'sum'}).sort_values('Revenue', ascending=False).head().InvoiceNo

22025    581483
2303     541431
18776    574941
19468    576365
9741     556444
Name: InvoiceNo, dtype: object

In [156]:
top5_revenue = retail.groupby('InvoiceNo', as_index=False).agg({'Revenue':'sum'}).sort_values('Revenue', ascending=False)

In [157]:
top5_revenue

InvoiceNo    Revenue
22025    581483  168469.60
2303     541431   77183.60
18776    574941   52940.94
19468    576365   50653.91
9741     556444   38970.00
...         ...        ...
25817   C580605  -17836.46
23338   C550456  -22998.40
23854   C556445  -38970.00
22510   C541433  -77183.60
25895   C581484 -168469.60

[25900 rows x 2 columns]

In [162]:
top5_revenue[:5].InvoiceNo.str.cat(sep=', ')

'581483, 541431, 574941, 576365, 556444'